system/HTTP/URIFactory.php

```diff
@@ -0,0 +1,248 @@
+<?php
+
+/**
+ * This file is part of CodeIgniter 4 framework.
+ *
+ * (c) CodeIgniter Foundation <admin@codeigniter.com>
+ *
+ * For the full copyright and license information, please view
+ * the LICENSE file that was distributed with this source code.
+ */
+
+namespace CodeIgniter\HTTP;
+
+use Config\App;
+
+class URIFactory
+{
+    /**
+     * @var array Superglobal SERVER array
+     */
+    private array $server;
+
+    /**
+     * @var array Superglobal GET array
+     */
+    private array $get;
+
+    private App $appConfig;
+
+    /**
+     * @param array $server Superglobal $_SERVER array
+     * @param array $get    Superglobal $_GET array
+     */
+    public function __construct(array &$server, array &$get, App $appConfig)
+    {
+        $this->server    = &$server;
+        $this->get       = &$get;
```

Comment:
I don't think passing by reference, which means modifying superglobal arrays, is the right approach.